### ISSUE TITLE
http client crashes (pthread lock bugs)

### ISSUE THREAD

**Issue by gdamore** [nanomsg/nng#1037]
I've seen this a few times in CI lately:

12/81 Test #12: httpclient .......................***Exception: Child aborted  0.36 sec
=== RUN: HTTP Client

  Given a TCP connection to example.com ✔✔✔✔
    We can initiate a message ✔✔✔✔✔✔
      The message contents are correct ✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔✔
  Given a client ✔✔✔
    One off exchange works ✔✔✔✔✔✔✔✔
    Timeout works ✔✔✔✔✔✔
    Connection reuse works ✔✔✔panic: pthread_mutex_lock: Invalid argument
This message is indicative of a BUG.
Report this at https://github.com/nanomsg/nng/issues
/home/runner/work/nng/nng/build/tests/httpclient(nni_panic+0x110) [0x55e6a5c4e1bb]
/home/runner/work/nng/nng/build/tests/httpclient(+0x341ee) [0x55e6a5c581ee]
/home/runner/work/nng/nng/build/tests/httpclient(nni_plat_mtx_lock+0x18) [0x55e6a5c5838e]
/home/runner/work/nng/nng/build/tests/httpclient(nni_mtx_lock+0x18) [0x55e6a5c554c1]
/home/runner/work/nng/nng/build/tests/httpclient(+0x5547b) [0x55e6a5c7947b]
/home/runner/work/nng/nng/build/tests/httpclient(+0x34fe0) [0x55e6a5c58fe0]
/home/runner/work/nng/nng/build/tests/httpclient(+0x31650) [0x55e6a5c55650]
/home/runner/work/nng/nng/build/tests/httpclient(+0x345cb) [0x55e6a5c585cb]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x76db) [0x7f9a756636db]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f9a7518488f]

**Comment by gdamore** [comment]
Here's another one from darwin:
```
  Given a client ✔✔✔
    One off exchange works ✔✔✔✔✔✔✔✔
    Timeout works ✔✔✔✔panic: pthread_mutex_lock: Invalid argument✔
This message is indicative of a BUG.
✔Report this at https://github.com/nanomsg/nng/issues

    Connection reuse works ✔✔✔1   httpclient                          0x00000001062bdb3c nni_panic + 524
2   httpclient                          0x00000001062d178a nni_pthread_mutex_lock + 58
3   httpclient                          0x00000001062d1745 nni_plat_mtx_lock + 21
4   httpclient                          0x00000001062c6ec5 nni_mtx_lock + 21
5   httpclient                          0x00000001062d00bd tcp_dialer_cb + 61
6   httpclient                          0x00000001062d2b5b nni_posix_poll_thr + 603
7   httpclient                          0x00000001062c71f8 nni_thr_wrap + 184
8   httpclient                          0x00000001062d1bd8 nni_plat_thr_main + 120
9   libsystem_pthread.dylib             0x00007fff686e6d36 _pthread_start + 125
10  libsystem_pthread.dylib             0x00007fff686e358f thread_start + 15
```

I've also seen a problem on Windows with this test, although no stack trace there.

**Comment by gdamore** [comment]
One possible thing that might be happening is that if we have threads still running in the background when atexit() calls nng_fini.  But I'd expect to see other tests besides this one fail in that case.

**Comment by gdamore** [comment]
It looks like probably that canceling the dialer is not unlinking it from pollfds, leading to a mutex panic as the dialer is later destroyed.

We need to be certain that pending dial operations are unregistered safely from the poller.

This is pretty platform specific, but Windows may have a similar logic failure.

**Comment by gdamore** [comment]
So I've done a bit more analysis -- the problem is that the dialer expects that closing and freeing the associated stream objects really does destroy them -- but that is a deferred deletion, and there might be a callback still pending on them.

I'm looking at options.  One option, which got some success, but I don't particularly like, is to have a way to drain the pollqs.  This requires changing pretty much every pollq implementation.  My first attempt at this reduced the problem incident rate *hugely*, but did not appear to solve it entirely -- I think likely there was another race in there.

The second option, which I intend to explore today if I have time, is to use a reference count on the dialer (and probably also the listener), for the streams it creates, and then use a kind of deferred destroy for them as well.  So they won't get truly deleted until the last reference on them is dropped.

This second approach should be rock solid, will require changing less code, and should also have the benefit of greatly reducing the extra locking that the first approach required.

**Comment by gdamore** [comment]
Looking at the Windows code, I don't think it suffers the same defect.  The io completion code is quite a bit different on Windows.

**Comment by gdamore** [comment]
The listen path is also clean.  IPC does have the same defect, but most of the time IPC connection attempts won't be backgrounded, so they are somewhat unlikely to see this, where as connect() from TCP over a long haul network is much more likely to run into it.   (Which is why httpclient is getting crushed by it -- it connects over a long network.)

This may also be responsible for crashes in websocket.
